vllogs: implement tenants endpoint for victorialogs

dmitryk-dk opened the pull request:
### Describe Your Changes

Please provide a brief description of the changes you made. Be as specific as possible to help others understand the purpose and impact of your modifications.

### Checklist

The following checks are **mandatory**:

- [ ] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

alongwill commented:
Let me know if you need help validating this. We're interested in using this feature! 🙏

zoutao0909 commented:
When will multi-tenant joint query be supported?

dmitryk-dk commented:
> When will multi-tenant joint query be supported?

Open a feature request or check the issues in the GitHub if it's present, and follow it

dmitryk-dk commented on lib/logstorage/storage_search.go:
```diff
@@ -944,6 +974,69 @@ func (s *Storage) search(workersCount int, so *genericSearchOptions, stopCh <-ch
 	}
 }
 
+func (s *Storage) searchPartitions(workersCount int, stopCh <-chan struct{}, process func(*partition, uint), start, end int64) {
```
logic was moved to the `getTenantIDs` function

dmitryk-dk commented on lib/logstorage/storage_search.go:
```diff
@@ -944,6 +974,69 @@ func (s *Storage) search(workersCount int, so *genericSearchOptions, stopCh <-ch
+			<-partitionSearchConcurrencyLimitCh
+		}(ptw.pt)
+	}
+	wgSearchers.Wait()
```
this part of the logic was simplified

dmitryk-dk commented:
@zekker6 after the changes I tested it. Please check the image below
<img width="830" alt="Screenshot 2025-04-01 at 13 13 30" src="https://github.com/user-attachments/assets/6ba9c85d-a760-4d1e-86e0-3c9a193c0aeb" />

dmitryk-dk commented:
Implemented logic for the cluster version. Updated response. Here is the response example
<img width="947" alt="Screenshot 2025-04-30 at 19 25 30" src="https://github.com/user-attachments/assets/14d7e3ba-2a3b-49bd-839f-2dbbd5eeec52" />

zekker6 commented on lib/logstorage/storage_search.go:
```diff
@@ -504,6 +505,92 @@ func (s *Storage) GetStreamIDs(ctx context.Context, tenantIDs []TenantID, q *Que
+		}
+	}
+
+	tenants := make([]TenantID, 0, len(m))
+	for tid := range m {
+		t, err := ParseTenantID(tid)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse tenantID %q: %w", tid, err)
+		}
+		tenants = append(tenants, t)
+	}
+	return json.Marshal(tenants)
```
Why parsing tenants list again?
Seems like it is only used in order to use `json.Marshal` for a list of tenants further. I would suggest to just return it in format of `accountID:projectID`. It is consistent with how it is handled in metrics and will require less parsing.

dmitryk-dk replied:
used the same behavior 
<img width="1660" alt="Screenshot 2025-05-09 at 14 40 02" src="https://github.com/user-attachments/assets/34168188-dd6e-4974-9992-46539b3553a6" />

zekker6 commented on app/vlselect/logsql/logsql.go:
```diff
@@ -938,6 +938,55 @@ func ProcessQueryRequest(ctx context.Context, w http.ResponseWriter, r *http.Req
+		return
+	}
+
+	_, err = fmt.Fprintf(w, "{\"status\":\"success\",\"data\":%q}\n", tenants)
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot obtain tenantIDs: %s", err)
+		return
+	}
```
Encoding `"data\":%q` data as quouted string causes it to be double escaped, for example:
```bash
curl localhost:9428/select/admin/tenants -d start=2022-07-13T12:16:51.229Z -d end=2025-07-12T12:16:51.229Z
{"status":"success","data":"[{\"accountID\":0,\"projectID\":0},{\"accountID\":10,\"projectID\":0},{\"accountID\":12,\"projectID\":11},{\"accountID\":10,\"projectID\":10},{\"accountID\":12,\"projectID\":1},{\"accountID\":12,\"projectID\":10}]"}
```
Parsed as JSON:
```json
{
  "status": "success",
  "data": "[{\"accountID\":12,\"projectID\":1},{\"accountID\":12,\"projectID\":10},{\"accountID\":12,\"projectID\":11},{\"accountID\":0,\"projectID\":0},{\"accountID\":10,\"projectID\":0},{\"accountID\":10,\"projectID\":10}]"
}
```

dmitryk-dk replied:
Fixed
<img width="1218" alt="Screenshot 2025-05-09 at 14 39 40" src="https://github.com/user-attachments/assets/a790817f-1ff5-4804-b918-7c7d94175f3c" />

zekker6 commented on app/vlselect/logsql/logsql.go:
```diff
@@ -938,6 +938,55 @@ func ProcessQueryRequest(ctx context.Context, w http.ResponseWriter, r *http.Req
 
+// ProcessAdminTenantsRequest processes /select/admin/tenants request.
+func ProcessAdminTenantsRequest(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	start, okStart, err := getTimeNsec(r, "start")
+	if err != nil {
+		httpserver.Errorf(w, r, "%s", err)
+		return
+	}
+	end, okEnd, err := getTimeNsec(r, "end")
+	if err != nil {
+		httpserver.Errorf(w, r, "%s", err)
+		return
+	}
+	if !okStart {
+		start = math.MinInt64
+	}
+	if !okEnd {
+		end = math.MaxInt64
+	}
```
I would suggest to add a validation that `start > end`. Otherwise an error in user input will be silently ignored and will return no results as all part wrappers will be filtered out.

dmitryk-dk replied:
added an additional check. Thank you for your review

vadimalekseev commented on app/vlstorage/netselect/netselect.go:
```diff
@@ -406,6 +418,48 @@ func (s *Storage) GetStreamIDs(ctx context.Context, tenantIDs []logstorage.Tenan
+
+	var tenantIDs []byte
+	for i := range results {
+		tenantIDs = append(tenantIDs, results[i]...)
```
Please, let me know if I made a mistake somewhere or if you need any additional information from me
